src/datetime_rotation.jl

```diff
@@ -0,0 +1,118 @@
+"""
+mutable struct DatetimeRotatingFileLogger <: AbstractLogger
+    logger::SimpleLogger
+    dir::String
+    filename_pattern::DateFormat
+    next_reopen_check::DateTime
+    always_flush::Bool
+end
+
+function DatetimeRotatingFileLogger(dir, filename_pattern; always_flush=true)
+    format = DateFormat(filename_pattern)
+    return DatetimeRotatingFileLogger(
+        SimpleLogger(open(calc_logpath(dir, filename_pattern), "a"), BelowMinLevel),
+        dir,
+        format,
+        next_datetime_transition(format),
+        always_flush,
+    )
+end
+
+function reopen!(drfl::DatetimeRotatingFileLogger)
+    drfl.logger = SimpleLogger(open(calc_logpath(drfl.dir, drfl.filename_pattern), "a"), BelowMinLevel)
+    drfl.next_reopen_check = next_datetime_transition(drfl.filename_pattern)
+    return nothing
+end
+
+# I kind of wish these were defined in Dates
+isless(::Type{Millisecond}, ::Type{Millisecond}) = false
+isless(::Type{Millisecond}, ::Type{T}) where {T <: Dates.Period} = true
+
+isless(::Type{Second}, ::Type{Millisecond}) = false
+isless(::Type{Second}, ::Type{Second}) = false
+isless(::Type{Second}, ::Type{T}) where {T <: Dates.Period} = true
+
+isless(::Type{Minute}, ::Type{Millisecond}) = false
+isless(::Type{Minute}, ::Type{Second}) = false
+isless(::Type{Minute}, ::Type{Minute}) = false
+isless(::Type{Minute}, ::Type{T}) where {T <: Dates.Period} = true
+
+isless(::Type{Hour}, ::Type{Day}) = true
+isless(::Type{Hour}, ::Type{Month}) = true
+isless(::Type{Hour}, ::Type{Year}) = true
+isless(::Type{Hour}, ::Type{T}) where {T <: Dates.Period} = false
+
+isless(::Type{Day}, ::Type{Month}) = true
+isless(::Type{Day}, ::Type{Year}) = true
+isless(::Type{Day}, ::Type{T}) where {T <: Dates.Period} = false
+
+isless(::Type{Month}, ::Type{Year}) = true
+isless(::Type{Month}, ::Type{T}) where {T <: Dates.Period} = false
+
+isless(::Type{Year}, ::Type{T}) where {T <: Dates.Period} = false
+
+"""
+    next_datetime_transition(fmt::DateFormat)
+
+Given a DateFormat that is being applied to our filename, what is the next
+time at which our filepath will need to change?
+"""
+function next_datetime_transition(fmt::DateFormat)
+    extract_token(x::Dates.DatePart{T}) where {T} = T
+    token_timescales = Dict(
+        # Milliseconds is the smallest timescale
+        's' => Millisecond,
+        # Seconds
+        'S' => Second,
+        # Minutes
+        'M' => Minute,
+        # Hours
+        'I' => Hour,
+        'H' => Hour,
+        # Days
+        'd' => Day,
+        'e' => Day,
+        'E' => Day,
+        # Month
+        'm' => Month,
+        'u' => Month,
+        'U' => Month,
+        # Year
+        'y' => Year,
+        'Y' => Year,
+    )
+
+    tokens = filter(t -> isa(t, Dates.DatePart), collect(fmt.tokens))
+    minimum_timescale = minimum(map(t -> token_timescales[extract_token(t)], tokens))
+    return Dates.ceil(now(), minimum_timescale) - Second(1)
```

Comment:
What if we used an  instance, rather than a type?
I think then we we can sort them without the type-piracy on `isless`
and i think rounding still works.

```suggestion
    minimum_timescale = minimum(map(t -> token_timescales[extract_token(t)](1), tokens))
    return ceil(now(), minimum_timescale) - Second(1)
```

Also `Dates.ceil` is just `Base.ceil`

Reply:
Also it is cool that with a small change we can make this to rotaed every 5 minutes etc since `ceil(now(), Minute(5))`

Reply:
How do you suggest I use instances like `Minute(5)`?  The whole point of this micjiggery is so that you can define a filename with a pattern embedded within it, and when that pattern is set to change, it will automatically roll the logs over for you.  I'm using the `DateFormat` type as a convenient "julian" way of specifying such a pattern, but I'm not married to that implementation at all.  I agree the type piracy is ugly.

Reply:
yeah, you can't actually use `Minute(5)` for that reason.

But you *can* use `Minute(1)` and `Hour(1)` etc.
And then `lessthan` just works, as does `ceil`

Reply:
Great; there are unfortunately a few cases where `isless()` doesn't work, but I've fixed that by just defining my own `custom_isless()` to avoid piracy.